### PR TITLE
[DRAFT] move enable sounds to the audio helper class and expose getter and setters to enable and disable toggeable sounds

### PR DESCRIPTION
**Pull request by kamalbennani**
The goal of this PR is to answer the "lack of implementation" described in this [issue](https://github.com/twilio/twilio-voice.js/issues/14)

**Contributing to Twilio**

> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### JIRA link(s):

- [CLIENT-0000](https://issues.corp.twilio.com/browse/CLIENT-0000)

### Description

This PR introduces a couple of changes:
- Remove enabledSounds responsibility from the Device class and move it close to the AudioHelper
- Expose toggleable sounds getters/setters
- Use `this.audio.outgoing` instead of relying on the enabledSounds private property

### Before review
* [ ] Updated CHANGELOG.md if necessary
* [ ] Added unit tests if necessary
* [ ] Updated affected documentation
* [ ] Verified locally with `npm test`
* [ ] Manually sanity tested running locally
* [ ] Ready for review

### Before merge
* [ ] Got one or more +1s
* [ ] Squashed erroneous commits if necessary
* [ ] Re-tested if necessary

### CODE REVIEW

**Comment by charliesantos** [comment]
@mhuynh5757 can you please do a quick pass?

File: lib/twilio/audiohelper.ts [hunk 1]
```diff
@@ -359,8 +387,8 @@ class AudioHelper extends EventEmitter {
       return enabledSounds[key];
     }
 
-    Object.keys(enabledSounds).forEach(key => {
-      (this as any)[key] = setValue.bind(null, key);
+    Object.keys(enabledSounds).forEach((key: Device.ToggleableSound) => {
```
Comment (mhuynh5757) [inline lib/twilio/audiohelper.ts:390]: 👍

File: lib/twilio/audiohelper.ts [hunk 2]
```diff
@@ -172,9 +201,7 @@ class AudioHelper extends EventEmitter {
     this.isOutputSelectionSupported = isEnumerationSupported && isSetSinkSupported;
     this.isVolumeSupported = isAudioContextSupported;
 
-    if (options.enabledSounds) {
-      this._addEnabledSounds(options.enabledSounds);
-    }
+    this._addEnabledSounds();
```
Comment (mhuynh5757) [inline lib/twilio/audiohelper.ts:204]: It seems like we only call this once at instantiation, and therefore we only add sounds that are enabled at construction time. It's not immediately clear to me that we can enable sounds post-construction, perhaps we need some mechanism for that?

File: lib/twilio/audiohelper.ts [hunk 3]
```diff
@@ -21,6 +21,12 @@ const kindAliases: Record<string, string> = {
   audiooutput: 'Audio Output',
 };
 
+export const defaultEnabledSounds: Record<Device.ToggleableSound, boolean> = {
```
Comment (mhuynh5757) [inline lib/twilio/audiohelper.ts:24]: I think we can remove these defaults from `device` if we want the `AudioHelper` class to have more control over the enablement of sounds.

Comment (mhuynh5757) [reply lib/twilio/audiohelper.ts:24]: Sorry, didn't see the `device` changes, please ignore this comment! 😓

File: lib/twilio/device.ts [hunk 4]
```diff
@@ -972,7 +963,7 @@ class Device extends EventEmitter {
       maxAverageBitrate: this._options.maxAverageBitrate,
       preflight: this._options.preflight,
       rtcConstraints: this._options.rtcConstraints,
-      shouldPlayDisconnect: () => this._enabledSounds.disconnect,
+      shouldPlayDisconnect: () => this._audio ? this._audio.disconnect() : defaultEnabledSounds.disconnect,
```
Comment (mhuynh5757) [inline lib/twilio/device.ts:966]: We should check if these members are defined, i.e. `this._audio.disconnect` might be undefined depending on the state of the `AudioHelper` and we might try to invoke it.

**Comment by mhuynh5757** [comment]
@kamalbennani thanks for contributing to the SDK! All feedback is greatly appreciated. I left some comments on some more specific details.

One recommendation over the current implementation that I'd like to see change is moving away from setting `EnableSoundFn` members on the object and using something less anti-pattern, like indexing an internal map with the `Sound` enumeration we already have. So the API would look something like ```
audioHelper.shouldPlaySound(Device.SoundName.Incoming)
```
or maybe we relegate all "shouldPlay" logic to the `AudioHelper` entirely and in the `Device` we just ask the `AudioHelper` to play it, getting some sort of return value from the `AudioHelper` if necessary.

What do you think?

**Comment by charliesantos** [comment]
@kamalbennani I'm closing this now. See my comment here https://github.com/twilio/twilio-voice.js/issues/14#issuecomment-1489266693

**Pull request closed without merging**